### PR TITLE
[CWS][SEC-6366] AD: drop event if its process lineage is incomplete

### PR DESCRIPTION
**Pull request by spikat**
### What does this PR do?

This PR adds two guards with dedicated new metrics:
- We now ensure that we did not add an event with a broken process context lineage.
- We also ensure to not send empty dumps.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/security/probe/activity_dump.go [hunk 1]
```diff
@@ -634,7 +648,7 @@ func (ad *ActivityDump) findOrCreateProcessActivityNode(entry *model.ProcessCach
 
 		// if none of them matched, create a new ProcessActivityNode for the input processCacheEntry
 		node = NewProcessActivityNode(entry, generationType, &ad.nodeStats)
-		// insert in the list of root entries
+		// insert in the list of childrens
```
Comment (paulcacheux) [inline pkg/security/probe/activity_dump.go:651]:
```suggestion
		// insert in the list of children
```

File: pkg/security/probe/activity_dump.go [hunk 2]
```diff
@@ -513,6 +517,11 @@ func (ad *ActivityDump) isEventTypeTraced(event *model.Event) bool {
 	return false
 }
 
+// IsEmpty return true if the dump did not contain any nodes
+func (ad *ActivityDump) IsEmpty() bool {
+	return len(ad.ProcessActivityTree) == 0
```
Comment (paulcacheux) [inline pkg/security/probe/activity_dump.go:522]: I think this should lock the mutex

Comment (spikat) [reply pkg/security/probe/activity_dump.go:524]: Suggestion applied on last [commit](https://github.com/DataDog/datadog-agent/pull/15013/commits/a68907931d767167f51519c00567f05042cf1baa), thank you

File: pkg/security/secl/model/process_cache_entry.go [hunk 3]
```diff
@@ -59,6 +59,21 @@ func (pc *ProcessCacheEntry) GetNextAncestorNoFork() *ProcessCacheEntry {
...
+		return true
+	} else if pc.Ancestor == nil {
+		return false
+	}
+
+	if maxDepth == 0 {
+		return false
+	}
+
+	return pc.Ancestor.HasCompleteLineage(maxDepth - 1)
```
Comment (lebauce) [inline pkg/security/secl/model/process_cache_entry.go:74]: nit:
maxDepth checks seems uncessary (as we never fail)
reuse this function in functional tests

Comment (spikat) [reply pkg/security/secl/model/process_cache_entry.go]: Indeed, failure is not an option. Guard removed

File: pkg/security/secl/model/process_cache_entry.go [hunk 4]
```diff
@@ -59,6 +59,16 @@ func (pc *ProcessCacheEntry) GetNextAncestorNoFork() *ProcessCacheEntry {
 	return ancestor
 }
 
+// HasCompleteLineage returns false if, from the entry, we cannot ascend the ancestors list to PID 1
+func (pc *ProcessCacheEntry) HasCompleteLineage() bool {
+	if pc.Pid == 1 { // we reach pid 1, the lineage is complete
```
Comment (safchain) [inline pkg/security/secl/model/process_cache_entry.go:64]: I think we could convert it to a non-recursive version ?

Comment (spikat) [reply pkg/security/secl/model/process_cache_entry.go]: Updated to non-recursive on last commit

**Review by safchain — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

**Pull request merged**